--- /nethome/dkulkarni47/.conda/envs/fifo-advisor/lib/python3.12/site-packages/lightningsim/runner.py.before-vitis2024-util-patch	2026-09-21 21:45:21.885744312 -0400
+++ /nethome/dkulkarni47/.conda/envs/fifo-advisor/lib/python3.12/site-packages/lightningsim/runner.py	2026-09-21 21:46:36.449943345 -0400
@@ -424,6 +424,31 @@
                         )
                     )
 
+                    util_input_path = (
+                        self.solution.path / f".autopilot/db/apatb_{kernel}_util.cpp"
+                    )
+                    util_path = tempdir / f"apatb_{kernel}_util.o"
+                    compile_util = create_task(
+                        run(
+                            [
+                                CC,
+                                "-std=c++14",
+                                "-I",
+                                xilinx_hls / "include",
+                                "-I",
+                                xilinx_hls / "lnx64/tools/systemc/include",
+                                "-I",
+                                SYSTEMC_ROOT / "include",
+                                "-c",
+                                util_input_path,
+                                "-g",
+                                "-fpermissive",
+                                "-o",
+                                util_path,
+                            ],
+                        )
+                    )
+
                     project_object_paths = [
                         tempdir / f"testbench.{i}.o"
                         for i, _ in enumerate(project_source_files)
@@ -502,6 +527,9 @@
                     compile_mapper_hw = await compile_mapper_hw
                     self.add_completed_process(compile_mapper_hw)
                     compile_mapper_hw.check_returncode()
+                    compile_util = await compile_util
+                    self.add_completed_process(compile_util)
+                    compile_util.check_returncode()
                     compile_project_files = [
                         compilation_process
                         for compilation_processes in await compile_project_files
@@ -512,7 +540,12 @@
                         compilation.check_returncode()
                     link_objects = " ".join(
                         str(path).replace(" ", "\\ ")
-                        for path in (*project_object_paths, object_path, mapper_hw_path)
+                        for path in (
+                            *project_object_paths,
+                            object_path,
+                            mapper_hw_path,
+                            util_path,
+                        )
                     )
                     extra_ld_flags = shlex.join(
                         str(flag)
